TestPurgeFileHoldingLockFile fails in semaphore

```
2016-07-17 06:59:58.489255 I | fileutil: purged file /tmp/purgefile255810807/0.test successfully
2016-07-17 06:59:58.489421 I | fileutil: purged file /tmp/purgefile255810807/1.test successfully
2016-07-17 06:59:58.489479 I | fileutil: purged file /tmp/purgefile255810807/2.test successfully
2016-07-17 06:59:58.502014 I | fileutil: purged file /tmp/purgefile255810807/3.test successfully
2016-07-17 06:59:58.516947 I | fileutil: purged file /tmp/purgefile255810807/4.test successfully
2016-07-17 06:59:58.526532 I | fileutil: purged file /tmp/purgefile255810807/5.test successfully
2016-07-17 06:59:58.541437 I | fileutil: purged file /tmp/purgefile255810807/6.test successfully
2016-07-17 06:59:58.550338 I | fileutil: purged file /tmp/purgefile439700458/0.test successfully
2016-07-17 06:59:58.550446 I | fileutil: purged file /tmp/purgefile439700458/1.test successfully
2016-07-17 06:59:58.550502 I | fileutil: purged file /tmp/purgefile439700458/2.test successfully
2016-07-17 06:59:58.550553 I | fileutil: purged file /tmp/purgefile439700458/3.test successfully
2016-07-17 06:59:58.550602 I | fileutil: purged file /tmp/purgefile439700458/4.test successfully
2016-07-17 06:59:58.566723 I | fileutil: purged file /tmp/purgefile439700458/5.test successfully
2016-07-17 06:59:58.566804 I | fileutil: purged file /tmp/purgefile439700458/6.test successfully
2016-07-17 06:59:58.678471 I | fileutil: purged file /tmp/purgefile242953684/0.test successfully
2016-07-17 06:59:58.678583 I | fileutil: purged file /tmp/purgefile242953684/1.test successfully
2016-07-17 06:59:58.678643 I | fileutil: purged file /tmp/purgefile242953684/2.test successfully
2016-07-17 06:59:58.689115 I | fileutil: purged file /tmp/purgefile242953684/3.test successfully
2016-07-17 06:59:58.699337 I | fileutil: purged file /tmp/purgefile242953684/4.test successfully
2016-07-17 06:59:58.709703 I | fileutil: purged file /tmp/purgefile242953684/5.test successfully
2016-07-17 06:59:58.723438 I | fileutil: purged file /tmp/purgefile242953684/6.test successfully
2016-07-17 06:59:58.734472 I | fileutil: purged file /tmp/purgefile259074851/0.test successfully
2016-07-17 06:59:58.734582 I | fileutil: purged file /tmp/purgefile259074851/1.test successfully
2016-07-17 06:59:58.734646 I | fileutil: purged file /tmp/purgefile259074851/2.test successfully
2016-07-17 06:59:58.734704 I | fileutil: purged file /tmp/purgefile259074851/3.test successfully
2016-07-17 06:59:58.734761 I | fileutil: purged file /tmp/purgefile259074851/4.test successfully
2016-07-17 06:59:58.745817 I | fileutil: purged file /tmp/purgefile259074851/5.test successfully
2016-07-17 06:59:58.745906 I | fileutil: purged file /tmp/purgefile259074851/6.test successfully
2016-07-17 06:59:58.863553 I | fileutil: purged file /tmp/purgefile523444373/0.test successfully
2016-07-17 06:59:58.863665 I | fileutil: purged file /tmp/purgefile523444373/1.test successfully
2016-07-17 06:59:58.863730 I | fileutil: purged file /tmp/purgefile523444373/2.test successfully
2016-07-17 06:59:58.883533 I | fileutil: purged file /tmp/purgefile523444373/3.test successfully
2016-07-17 06:59:58.916460 I | fileutil: purged file /tmp/purgefile523444373/4.test successfully
2016-07-17 06:59:58.925566 I | fileutil: purged file /tmp/purgefile523444373/5.test successfully
2016-07-17 06:59:58.936863 I | fileutil: purged file /tmp/purgefile523444373/6.test successfully
2016-07-17 06:59:58.947136 I | fileutil: purged file /tmp/purgefile977304816/0.test successfully
2016-07-17 06:59:59.070026 I | fileutil: purged file /tmp/purgefile977304816/1.test successfully
2016-07-17 06:59:59.070117 I | fileutil: purged file /tmp/purgefile977304816/2.test successfully
2016-07-17 06:59:59.070187 I | fileutil: purged file /tmp/purgefile977304816/3.test successfully
2016-07-17 06:59:59.070253 I | fileutil: purged file /tmp/purgefile977304816/4.test successfully
2016-07-17 06:59:59.070317 I | fileutil: purged file /tmp/purgefile977304816/5.test successfully
2016-07-17 06:59:59.070383 I | fileutil: purged file /tmp/purgefile977304816/6.test successfully
--- FAIL: TestPurgeFileHoldingLockFile (0.13s)
	purge_test.go:126: filenames = [1.test 2.test 3.test 4.test 5.test 6.test 7.test 8.test 9.test], want [5.test 6.test 7.test 8.test 9.test]
FAIL
coverage: 58.7% of statements
```

--- a/pkg/fileutil/purge.go
+++ b/pkg/fileutil/purge.go
@@ -23,6 +23,11 @@ import (
 )
 
 func PurgeFile(dirname string, suffix string, max uint, interval time.Duration, stop <-chan struct{}) <-chan error {
+	return purgeFile(dirname, suffix, max, interval, stop, nil)
+}
+
+// purgeFile is the internal implementation for PurgeFile which can post purged files to purgec if non-nil.
+func purgeFile(dirname string, suffix string, max uint, interval time.Duration, stop <-chan struct{}, purgec chan<- string) <-chan error {
 	errC := make(chan error, 1)
 	go func() {
 		for {
@@ -38,6 +43,7 @@ func PurgeFile(dirname string, suffix string, max uint, interval time.Duration, 
 				}
 			}
 			sort.Strings(newfnames)
+			fnames = newfnames
 			for len(newfnames) > int(max) {
 				f := path.Join(dirname, newfnames[0])
 				l, err := TryLockFile(f, os.O_WRONLY, PrivateFileMode)
@@ -55,6 +61,11 @@ func PurgeFile(dirname string, suffix string, max uint, interval time.Duration, 
 				}
 				plog.Infof("purged file %s successfully", f)
 				newfnames = newfnames[1:]
+			}
+			if purgec != nil {
+				for i := 0; i < len(fnames)-len(newfnames); i++ {
+					purgec <- fnames[i]
+				}
 			}
 			select {
 			case <-time.After(interval):

--- a/pkg/fileutil/purge_test.go
+++ b/pkg/fileutil/purge_test.go
@@ -31,44 +31,48 @@ func TestPurgeFile(t *testing.T) {
 	}
 	defer os.RemoveAll(dir)
 
-	for i := 0; i < 5; i++ {
-		var f *os.File
-		f, err = os.Create(path.Join(dir, fmt.Sprintf("%d.test", i)))
-		if err != nil {
+	// minimal file set
+	for i := 0; i < 3; i++ {
+		f, ferr := os.Create(path.Join(dir, fmt.Sprintf("%d.test", i)))
+		if ferr != nil {
 			t.Fatal(err)
 		}
 		f.Close()
 	}
 
-	stop := make(chan struct{})
+	stop, purgec := make(chan struct{}), make(chan string, 10)
 
-	// keep at most 3 most recent files
-	errch := PurgeFile(dir, "test", 3, time.Millisecond, stop)
-
-	// create 5 more files
-	for i := 5; i < 10; i++ {
-		var f *os.File
-		f, err = os.Create(path.Join(dir, fmt.Sprintf("%d.test", i)))
-		if err != nil {
-			t.Fatal(err)
-		}
-		f.Close()
-		time.Sleep(10 * time.Millisecond)
+	// keep 3 most recent files
+	errch := purgeFile(dir, "test", 3, time.Millisecond, stop, purgec)
+	select {
+	case f := <-purgec:
+		t.Errorf("unexpected purge on %q", f)
+	case <-time.After(10 * time.Millisecond):
 	}
 
-	// purge routine should purge 7 out of 10 files and only keep the
-	// 3 most recent ones.
-	// Wait for purging for at most 300ms.
-	var fnames []string
-	for i := 0; i < 30; i++ {
-		fnames, err = ReadDir(dir)
-		if err != nil {
-			t.Fatal(err)
+	// rest of the files
+	for i := 4; i < 10; i++ {
+		go func(n int) {
+			f, ferr := os.Create(path.Join(dir, fmt.Sprintf("%d.test", n)))
+			if ferr != nil {
+				t.Fatal(err)
+			}
+			f.Close()
+		}(i)
+	}
+
+	// watch files purge away
+	for i := 4; i < 10; i++ {
+		select {
+		case <-purgec:
+		case <-time.After(time.Second):
+			t.Errorf("purge took too long")
 		}
-		if len(fnames) <= 3 {
-			break
-		}
-		time.Sleep(10 * time.Millisecond)
+	}
+
+	fnames, rerr := ReadDir(dir)
+	if rerr != nil {
+		t.Fatal(rerr)
 	}
 	wnames := []string{"7.test", "8.test", "9.test"}
 	if !reflect.DeepEqual(fnames, wnames) {
@@ -77,9 +81,11 @@ func TestPurgeFile(t *testing.T) {
 
 	// no error should be reported from purge routine
 	select {
+	case f := <-purgec:
+		t.Errorf("unexpected purge on %q", f)
 	case err := <-errch:
 		t.Errorf("unexpected purge error %v", err)
-	case <-time.After(time.Millisecond):
+	case <-time.After(10 * time.Millisecond):
 	}
 	close(stop)
 }
@@ -107,29 +113,33 @@ func TestPurgeFileHoldingLockFile(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	stop := make(chan struct{})
-	errch := PurgeFile(dir, "test", 3, time.Millisecond, stop)
+	stop, purgec := make(chan struct{}), make(chan string, 10)
+	errch := purgeFile(dir, "test", 3, time.Millisecond, stop, purgec)
 
-	var fnames []string
-	for i := 0; i < 10; i++ {
-		fnames, err = ReadDir(dir)
-		if err != nil {
-			t.Fatal(err)
+	for i := 0; i < 5; i++ {
+		select {
+		case <-purgec:
+		case <-time.After(time.Second):
+			t.Fatalf("purge took too long")
 		}
-		if len(fnames) <= 5 {
-			break
-		}
-		time.Sleep(10 * time.Millisecond)
 	}
+
+	fnames, rerr := ReadDir(dir)
+	if rerr != nil {
+		t.Fatal(rerr)
+	}
+
 	wnames := []string{"5.test", "6.test", "7.test", "8.test", "9.test"}
 	if !reflect.DeepEqual(fnames, wnames) {
 		t.Errorf("filenames = %v, want %v", fnames, wnames)
 	}
 
 	select {
+	case s := <-purgec:
+		t.Errorf("unexpected purge %q", s)
 	case err = <-errch:
 		t.Errorf("unexpected purge error %v", err)
-	case <-time.After(time.Millisecond):
+	case <-time.After(10 * time.Millisecond):
 	}
 
 	// remove the purge barrier
@@ -137,15 +147,18 @@ func TestPurgeFileHoldingLockFile(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	for i := 0; i < 10; i++ {
-		fnames, err = ReadDir(dir)
-		if err != nil {
-			t.Fatal(err)
+	// wait for rest of purges (5, 6)
+	for i := 0; i < 2; i++ {
+		select {
+		case <-purgec:
+		case <-time.After(time.Second):
+			t.Fatalf("purge took too long")
 		}
-		if len(fnames) <= 3 {
-			break
-		}
-		time.Sleep(10 * time.Millisecond)
+	}
+
+	fnames, rerr = ReadDir(dir)
+	if rerr != nil {
+		t.Fatal(rerr)
 	}
 	wnames = []string{"7.test", "8.test", "9.test"}
 	if !reflect.DeepEqual(fnames, wnames) {
@@ -153,9 +166,11 @@ func TestPurgeFileHoldingLockFile(t *testing.T) {
 	}
 
 	select {
+	case f := <-purgec:
+		t.Errorf("unexpected purge on %q", f)
 	case err := <-errch:
 		t.Errorf("unexpected purge error %v", err)
-	case <-time.After(time.Millisecond):
+	case <-time.After(10 * time.Millisecond):
 	}
 
 	close(stop)

